cpp/src/arrow/util/converter.h

```diff
@@ -302,32 +305,59 @@ class Chunker {
     return status;
   }
 
-  // we could get bit smarter here since the whole batch of appendable values
-  // will be rejected if a capacity error is raised
-  Status Extend(InputType values, int64_t size) {
-    auto status = converter_->Extend(values, size);
-    if (ARROW_PREDICT_FALSE(status.IsCapacityError())) {
-      if (converter_->builder()->length() == 0) {
+  Status Extend(InputType values, int64_t size, int64_t offset = 0) {
+    while (offset < size) {
+      auto length_before = converter_->builder()->length();
+      auto status = converter_->Extend(values, size, offset);
+      auto length_after = converter_->builder()->length();
+      auto num_converted = length_after - length_before;
+
+      offset += num_converted;
+      length_ += num_converted;
+
+      if (status.IsCapacityError()) {
+        if (converter_->builder()->length() == 0) {
+          // Builder length == 0 means the individual element is too large to append.
+          // In this case, no need to try again.
+          return status;
+        } else if (converter_->rewind_on_capacity_error()) {
```

Comment:
The list-like and binary-like conversion paths may raise capacity error, but there is a difference. 

While the binary-like converters check the capacity before append/extend the list-like converters first append/extend the value(s) and checks just afterwards. Thus depending on the implementation semantics we may need to rewind (slice) the output chunk by one.

Reply:
Can we document this inline? I was going to question this since it's rather confusing looking at it.

Reply:
Yes. I'm going to add comments based on these discussions.